File: backend/pkg/engine/rng.go
// Package engine implements the deterministic Vault Rush puzzle engine.
//
// The engine is intentionally free of floating point in every code path that
// influences board state so that the TypeScript port (client/src/engine) can
// reproduce it bit-for-bit. See docs/engine.md for the normative description.
package engine

// RNG is a mulberry32 generator. It is tiny, fast, and trivially portable to
// JavaScript (Math.imul + >>> 0), which is why it was chosen over Go's
// math/rand sources.
type RNG struct {
	state uint32
}

// NewRNG seeds a generator. Only the low 32 bits of seed are used.
func NewRNG(seed uint32) *RNG {
	return &RNG{state: seed}
}

// Next returns the next 32-bit value.
func (r *RNG) Next() uint32 {
	r.state += 0x6D2B79F5
	t := r.state
	t = (t ^ (t >> 15)) * (1 | t)
	t = (t + ((t ^ (t >> 7)) * (61 | t))) ^ t
	return t ^ (t >> 14)
}

// NextInt returns a value in [0, n). n must be > 0.
func (r *RNG) NextInt(n int) int {
	if n <= 0 {
		panic("engine: NextInt with n <= 0")
	}
	return int(r.Next() % uint32(n))
}

// State exposes the internal state for snapshot/debug purposes.
func (r *RNG) State() uint32 { return r.state }

File: client/src/engine/rng.ts
/**
 * mulberry32 — bit-for-bit identical to backend/pkg/engine/rng.go.
 * All arithmetic is forced to unsigned 32-bit; never derive floats from it.
 */
export class RNG {
  private s: number;

  constructor(seed: number) {
    this.s = seed >>> 0;
  }

  next(): number {
    this.s = (this.s + 0x6d2b79f5) >>> 0;
    let t = this.s;
    t = Math.imul(t ^ (t >>> 15), 1 | t) >>> 0;
    t = ((t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t) >>> 0;
    return (t ^ (t >>> 14)) >>> 0;
  }

  nextInt(n: number): number {
    if (n <= 0) throw new Error('engine: nextInt with n <= 0');
    return this.next() % n;
  }

  get state(): number {
    return this.s;
  }
}
